app/services/cache_manager.rb

```diff
@@ -10,7 +10,39 @@ def self.create_new_permission_cache
   end
 
   def self.create_new_filter_cache
-    Filter.all.each(&:save!)
+    locked_roles = Role.select(&:locked?)
+    origins = locked_roles.map(&:origin)
+    locked_roles.each { |role| role.origin = nil }
+    locked_roles.each { |role| role.modify_locked = true }
+    locked_roles.each(&:save)
+    Filter.all.each do |filter|
+      next if filter.save
+
+      if filter.errors[:permissions] == ['Permissions must be of same resource type']
+        # If the filter was created long time ago, it may have had several
+        # resources of the same type. To fix it, let's split it into different
+        # filters each with its own resource type.
```

Comment:
Is it possible to do this in a migration? If I understand correctly this is caused by some old filters that weren't migrated correctly and would only need to be done once. I wouldn't expect this to be part of the cache rebuild.